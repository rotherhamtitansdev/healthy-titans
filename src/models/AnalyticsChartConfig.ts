import { NHSNutritionalDataModel } from "./NHSNutritionalDataModel";

// This is the numerical instance of the NHS Nutritional guidelines model
export const NHSNutritionalBoundaries: NHSNutritionalDataModel[] = [
  { name: "Fat", low: 3, high: 17.5 },
  { name: "Saturates", low: 1.5, high: 5 },
  { name: "Sugars", low: 5, high: 22.5 },
  { name: "Salt", low: 0.1, high: 0.6 },
];

// These are the config parameters for the chart styling and sizing
export const chartSizing = {
  chartHeight: 500,
  barWidth: 30,
  CStartPos: 50,
  RStartPos: 60,
  circleYDif: 100,
  smallerRounded: 15,
};

export enum ChartLabel {
  Calories = "ENERC_KCAL",
  Protein = "PROCNT",
  Fat = "FAT",
  Saturates = "FASAT",
  Sugars = "SUGAR",
  Salt = "NA",
}

export enum Colour {
  Green = 1,
  Yellow = 2,
  Red = 3,
}
