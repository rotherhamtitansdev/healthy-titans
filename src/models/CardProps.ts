export interface CardProps {
  name: string;
  additionalStyling?: string;
  onClick?: () => void;
}
