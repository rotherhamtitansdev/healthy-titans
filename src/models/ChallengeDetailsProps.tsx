export interface ChallengeDetailsProps {
  name: string;
  description: string;
  img: string;
  rules: string[];
  equipment: string[];
  howToPlay?: string[];
}
