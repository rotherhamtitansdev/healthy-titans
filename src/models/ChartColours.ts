export interface ChartColours {
  name: string;
  primary: string;
  secondary: string;
  status: number;
  position: number;
}
