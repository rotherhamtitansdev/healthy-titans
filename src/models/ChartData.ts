export interface ChartData {
  Calories: number;
  Protein: number;
  Fat: number;
  Saturates: number;
  Sugars: number;
  Salt: number;
}