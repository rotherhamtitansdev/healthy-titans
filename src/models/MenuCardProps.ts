export interface MenuCardProps {
  key: number;
  name: string;
  description?: string;
  img?: string;
  path: string;
  externalPath?: boolean;
  disableOnClick?: boolean;
  firebaseName?: string;
  additionalStyling?: string;
}
