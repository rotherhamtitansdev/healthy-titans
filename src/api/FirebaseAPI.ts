import { getDownloadURL, getStorage, ref } from "firebase/storage";
import {
  collection,
  doc,
  getDoc,
  getDocs,
  query,
  where,
  documentId,
} from "firebase/firestore";
import { FoodDetailsProps } from "../data/nutritional_information/FoodDetailsComponentData";
import { fStore } from "../config/firebase-config";
import { NutritionalDetailsFirebaseProps } from "../models/NutritionDetailsComponentData";

interface SeeNextProps {
  key: number;
  name: string;
  path: string;
  img: string;
}

class FirebaseAPI {
  static fetchImages = async (firebaseName: string): Promise<string> => {
    const storage = getStorage();
    const starsRef = ref(storage, firebaseName);
    return getDownloadURL(starsRef);
  };

  static fetchAllImages = async (firebaseNames: string[]) =>
    Promise.all(firebaseNames.map((firebaseName) => FirebaseAPI.fetchImages(firebaseName)));

  static fetchFoodDetailsComponentsData = async (): Promise<
    FoodDetailsProps[] | undefined
  > => {
    const querySnapshot = await getDocs(collection(fStore, "FYPData"));
    if (!querySnapshot) return undefined;
     return querySnapshot.docs.map((each) => each.data() as FoodDetailsProps);
  };

  static fetchFoodDetailsSingle = async (name: string) => {
    const docRef = doc(fStore, "FYPData", name);
    const docSnap = await getDoc(docRef);
    if (docSnap) return docSnap.data() as FoodDetailsProps;
    return undefined;
  };

  static fetchFoodDetailsSeeNext = async (
    category: string,
    currentName: string
  ): Promise<{ cardData: FoodDetailsProps[]; paths: string[] }> => {
    const q = query(
      collection(fStore, "FYPData"),
      where("category", "==", category),
      where("name", "!=", currentName)
    );
    const docSnap = await getDocs(q);
    const docs = docSnap.docs.map((data) => {
      const newDoc = data.data();
      newDoc.path = data.id;
      return newDoc;
    });

    const shuffled = docs.sort(() => 0.5 - Math.random()).slice(0, 3);

    return {
      cardData: shuffled.map((value) => ({
        name: value.name,
        description: value.description,
        firebaseName: value.firebaseName,
        category: value.category,
        score: value.score,
        facts: value.facts,
      })),
      paths: shuffled.map((value) => value.path),
    };
  };

  static fetchNutritionData = async (name: string): Promise<NutritionalDetailsFirebaseProps> => {
    const querySnapshot = await getDocs(collection(fStore, "NutritionData", name, "Content"));

    const sorted = querySnapshot.docs.sort((a, b) => {
      if (a.data().order > b.data().order) {
        return 1;
      }
      if (b.data().order > a.data().order) {
        return -1;
      }
      return 0;
    });

    return sorted.map((doc2) => {
      const data = doc2.data();
      return Object.entries(data).map(([key,value]) => ({key,value}));
    });
  };

  static fetchNutritionSeeNext = async (currentName: string): Promise<SeeNextProps[]> => {
    const q = query(collection(fStore, "NutritionData"), where(documentId(), "!=", currentName));
    const querySnapshot = await getDocs(q);

    const docs = querySnapshot.docs.sort(() => 0.5 - Math.random()).slice(0, 3);

    return Promise.all(
      docs.map(async (sortedDoc, index) => {
        const dataDoc = await getDoc(doc(fStore, "NutritionData", sortedDoc.id, "Content", "Main"));

        const URI = await FirebaseAPI.fetchImages(dataDoc.data()?.firebaseName);

        return { key: index, name: dataDoc.data()?.name, path: docs[index].id, img: URI };
      })
    );
  };

  static fetchQuizData = async () => {
    const docRef = doc(fStore, "QuizData", "Quiz");
    const quizDoc = await getDoc(docRef);

    if (quizDoc.exists()) {
      return quizDoc.data();
    }
    return undefined;
  };

  static fetchAboutUsData = async () => {
    const docRef = doc(fStore, "AboutUsData", "Data");
    const aboutUsDoc = await getDoc(docRef);

    if (aboutUsDoc.exists()) {
      return aboutUsDoc.data();
    }
    return undefined;
  };

  static fetchChallengesData = async (currentName: string) => {
    const docRef = doc(fStore, "Challenges", currentName);
    const challengesDoc = await getDoc(docRef);

    if (challengesDoc.exists()) {
      return challengesDoc.data();
    }
    return undefined;
  };

  static fetchChallengesList = async () => {
    const querySnapshot = await getDocs(collection(fStore, "ChallengesData"));

    if (!querySnapshot) return undefined;
    return querySnapshot.docs.map((each) => each.data());
  };
}

export default FirebaseAPI;
