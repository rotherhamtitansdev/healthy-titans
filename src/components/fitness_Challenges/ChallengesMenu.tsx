import React, {useEffect, useState} from "react";
import { HeaderProps } from "../../models/HeaderProps";
import { MenuCardProps } from "../../models/MenuCardProps";
import Menu from "../shared/Menu";
import FirebaseAPI from "../../api/FirebaseAPI";


const Challenges = () => {
  const [data, setData] = useState<MenuCardProps[]>([])
  
  useEffect(()=>{
    FirebaseAPI.fetchChallengesList().then((res)=>setData(res as MenuCardProps[]))
  }, [])
  
    const headerData: HeaderProps = {
    title: "Fitness Challenges",
    body: "Being active and moving your body is really important to remain strong and healthy. Exercise can be in the form of playing a sport, walking the dog or taking part in competetive games with friends!",
  };

  return (
    <Menu
      header={headerData}
      cards={data}
      title={{
        title: "Stay fit with these fun challenges!",
        subtitle: "Click on a picture to learn more about each activity",
      }}
    />
  );
};

export default Challenges;
