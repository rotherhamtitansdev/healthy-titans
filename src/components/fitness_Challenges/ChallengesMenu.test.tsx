import React from "react";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import FirebaseAPI from "../../api/FirebaseAPI";
import ChallengesMenu from "./ChallengesMenu";
import { MenuCardProps } from "../../models/MenuCardProps";

const content: MenuCardProps[] = [
  {
    key: 0,
    name: "Ball Rush",
    path: "/BallRush",
    firebaseName: "FitnessChallengesImages/ballRush.svg",
  },
  {
    key: 1,
    name: "Ball Balance",
    path: "/BallBalance",
    firebaseName: "FitnessChallengesImages/ballBalance.svg",
  },
];

describe("ChallengesMenu", () => {
  test("component renders with list of challenges", async () => {
    jest.spyOn(FirebaseAPI, "fetchChallengesList").mockResolvedValue(content);

    const { asFragment } = render(<ChallengesMenu />, {
      wrapper: MemoryRouter,
    });

    expect(await screen.findByText("Ball Rush")).toBeVisible();
    expect(await screen.findByText("Ball Balance")).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });
});
