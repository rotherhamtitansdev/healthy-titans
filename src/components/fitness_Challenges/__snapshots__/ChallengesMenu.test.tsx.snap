// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChallengesMenu component renders with list of challenges 1`] = `
<DocumentFragment>
  <div
    class="bg-mobileWave sm:bg-wave bg-no-repeat pb-2 bg-span"
  >
    <div
      class="flex flex-wrap h-[22rem]"
    >
      <div
        class="mx-auto pt-[3rem] md:pt-[2rem]"
      >
        <h1
          class="text-titansDarkBlue font-bold text-center sm:mt-[1rem] text-3xl sm:text-5xl lg:text-6xl mt-8 pb-4"
          data-testid="header-title"
        >
          Fitness Challenges
        </h1>
        <div
          class="font-medium text-homepageHeaderSubTitle text-sm md:text-base lg:text-xl text-center w-2/3 lg:w-[45rem] mx-auto"
        >
          Being active and moving your body is really important to remain strong and healthy. Exercise can be in the form of playing a sport, walking the dog or taking part in competetive games with friends!
        </div>
      </div>
    </div>
  </div>
  <div
    class="p-5 md:p-10 pb-20"
  >
    <div
      class="md:px-5"
    >
      <div
        class="font-semibold text-[#243A73] text-2xl md:text-3xl lg:text-4xl py-5"
        data-testid="menutitle-title"
      >
        Stay fit with these fun challenges!
      </div>
      <div
        class="font-medium text-[#243A73] md:text-xl lg:text-xl md:pb-5"
      >
        Click on a picture to learn more about each activity
      </div>
    </div>
    <div
      class="grid grid-cols-2 xs:p-4 lg:grid-cols-auto lg:px-14 xl:px-20 2xl:px-30 3xl:px-36 lg:py-10 gap-[0.5rem] md:gap-y-[1rem] lg:gap-y-[3rem]"
      data-testid="menu-styling"
    >
      <div
        class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg lg:w-[17rem] xl:w-[21rem] xl:h-[21rem] 2xl:w-[25rem] 2xl:h-[25rem] hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
        data-testid="Ball Rush"
        role="button"
        tabindex="0"
      >
        <div
          class="py-[1.75rem] md:py-[3.625rem] w-full h-full flex-col flex"
        >
          <div
            class="w-full h-full flex justify-center items-center md:-mt-0 md:-mt-8"
          >
            <span
              class=""
              style="width: 100px; height: 100px; border-radius: 100%; border: 2px solid; border-top-color: #DC476D; border-bottom-color: transparent; border-left-color: #DC476D; border-right-color: #DC476D; display: inline-block; animation: react-spinners-ClipLoader-clip 0.75s 0s infinite linear; animation-fill-mode: both;"
            />
          </div>
          <div
            class="md:text-xl text-homepageHeaderText font-semibold text-center mt-auto"
          >
            Ball Rush
          </div>
        </div>
      </div>
      <div
        class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg lg:w-[17rem] xl:w-[21rem] xl:h-[21rem] 2xl:w-[25rem] 2xl:h-[25rem] hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
        data-testid="Ball Balance"
        role="button"
        tabindex="0"
      >
        <div
          class="py-[1.75rem] md:py-[3.625rem] w-full h-full flex-col flex"
        >
          <div
            class="w-full h-full flex justify-center items-center md:-mt-0 md:-mt-8"
          >
            <span
              class=""
              style="width: 100px; height: 100px; border-radius: 100%; border: 2px solid; border-top-color: #DC476D; border-bottom-color: transparent; border-left-color: #DC476D; border-right-color: #DC476D; display: inline-block; animation: react-spinners-ClipLoader-clip 0.75s 0s infinite linear; animation-fill-mode: both;"
            />
          </div>
          <div
            class="md:text-xl text-homepageHeaderText font-semibold text-center mt-auto"
          >
            Ball Balance
          </div>
        </div>
      </div>
      <div
        class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg lg:w-[17rem] xl:w-[21rem] xl:h-[21rem] 2xl:w-[25rem] 2xl:h-[25rem] col-span-2 mx-[25%] lg:col-span-1 lg:mx-0 hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
        role="button"
        tabindex="0"
      >
        <div
          class="py-[1.75rem] md:py-[3.625rem] w-full h-full flex-col flex"
        >
          <div
            class="w-full h-full flex justify-center items-center md:-mt-0 md:-mt-8"
          >
            <span
              class=""
              style="width: 100px; height: 100px; border-radius: 100%; border: 2px solid; border-top-color: #DC476D; border-bottom-color: transparent; border-left-color: #DC476D; border-right-color: #DC476D; display: inline-block; animation: react-spinners-ClipLoader-clip 0.75s 0s infinite linear; animation-fill-mode: both;"
            />
          </div>
          <div
            class="md:text-xl text-homepageHeaderText font-semibold text-center mt-auto"
          />
        </div>
      </div>
      <div
        class="grid col-span-2 lg:col-span-3 "
      />
    </div>
  </div>
</DocumentFragment>
`;
